### ISSUE TITLE
cabal new-haddock and project index

### ISSUE THREAD

**Issue by yav** [haskell/cabal#5268]
Hello,

I can't seem to find a way to tell cabal new-haddock to put the top-level index for a package in an index common for my project (e.g., something like haddock's `use-index` option).   Is this possible?  If not, this would be a handy feature to have.

**Comment by harpocrates** [comment]
You should be able to thread options directly to Haddock via `--haddock-options`.

    cabal new-haddock --haddock-options '--use-index=myindex.html'

**Comment by yav** [comment]
OK, @harpocrates and I figured out how to get this to work using the `haddock-options`, but it really would be nice if this just worked with `new-haddock` directly.  For example, we could have a flag that indicates if you want each package to have its own contents and index, or if there should be for the project, and all documentation should be joined together.

**Comment by 23Skidoo** [comment]
Would be nice if `new-haddock` did this out of the box.

**Comment by yav** [comment]
Indeed.  Currently I have a small script that does that, but there are a few problems with it, that wouldn't happen if `haddock` did it out of the box.  In particular:
  * since I have to pass extra flags, `cabal` ends up rebuilding a bunch of stuff, because the flags have changed, even though the flags are just about haddock, and have nothing to do with the state of the executable
   * there is a bunch of hackery I have to do to support switching versions of GHC.

**Comment by 23Skidoo** [comment]
@merijn's recent work (see e.g. #5287) should help with the former.

**Comment by merijn** [comment]
HEAD should already alleviate the worst problems, since non-local dependencies don't get rebuild on spurious flag changes. #5287 is just making sure the same goes for local/in-place dependencies.

**Comment by david-christiansen** [comment]
It would also be nice if there were a command to open the project's Haddock index in a Web browser. `raco docs` does this for Racket, and it's very convenient.

**Comment by harpocrates** [comment]
I started doing this last night https://github.com/harpocrates/cabal/commit/5a92825449c3bf9219f9bb24f5606c11fb54712f, but ended up giving up. In particular: it looks like the per-package information needed to make the right call to `haddock` (where `.haddock` interface files are and where the HTML output is) is tied up inside an `InstalledPackageInfo`. I'm having a great deal of difficulty finding all of the needed `InstalledPackageInfo`s. Some broken cases:

  * a package that is new-haddock'ed before it is new-built
  * an executable

Technical questions aside, there is also the question of which packages to include in the index (do we want just the packages on which we called `new-haddock`? do we want to include direct dependencies? etc.).

**Comment by frisbynavican** [comment]
@yav Would you share your workaround "script" here? Ticket seems dormant. Thanks!

**Comment by yav** [comment]
Yep, that's pretty much what I did, with just some extra hackery to deal with different GHC versions.  The script is here:

https://raw.githubusercontent.com/GaloisInc/crucible/master/scripts/gen-doc
